src/core/lwt.ml

```diff
@@ -22,1307 +22,2746 @@
+   able to make it!
+
+
+   Enjoy! *)
+
+
+
+(* Overview
+
+   In this file, there is a "model" function -- [Lwt.bind] -- which pulls
+   together many (though not all) of the concepts and helpers discussed in this
+   overview. To find it, search for "let bind," and you can examine it while
+   reading the overview. The authors of this file intend to put extra effort
+   into writing nice comments inside [Lwt.bind] :)
+
+
+   0. Main mechanism and two aspects
+
+   The Lwt interface ([lwt.mli]) provides one main mechanism, promises, and two
+   "aspects," which are *not* necessary to understand the main mechanism
+   promises, but they are still there:
+
+   - promise cancelation
+   - sequence-associated storage
+
+   If you are not interested in cancelation or storage, you can ignore these two
+   complications, and still get a pretty good understanding of the code. To
+   help, all identifiers related to cancelation contain the string "cancel," and
+   all identifiers related to storage contain "storage."
+
+
+   1. Promises
+
+   A promise is a cell that can be in one of two states: "completed" or
+   "pending."
+
+   - Completed promises
+
+     A completed promise is either "resolved" with a value, or "failed" with an
+     exception. The state of a completed promise will never change again: a
+     completed promise is immutable. A completed promise is basically equivalent
+     to an [('a, exn) Pervasives.result]. Completed promises are produced in two
+     ways:
+
+     - [Lwt.return], [Lwt.fail], and related functions, produce "trivial"
+       promises that are completed from the start.
+     - The other way is to complete a promise that started out pending.
+
+     Note that failed promises have nothing to do with unhandled exceptions.
+
+   - Pending promises
+
+     ...are those that may become completed in the future. Each pending promise
+     carries a list of callbacks. These callbacks are added by functions like
+     [Lwt.bind], and called by Lwt if/when the promise completes. Pending
+     promises are produced in three ways, according to how they can be
+     completed:
+
+     - Initial promises
+
+       ...are created by [Lwt.wait] and [Lwt.task]. The user of Lwt completes
+       these promises manually, through the resolvers returned by those
+       functions.
+
+     - Sequential composition
+
+       For example, [Lwt.bind]. These promises only complete when some sequence
+       of "preceding" promises completes. The user cannot complete these
+       promises directly.
+
+     - Concurrent composition
+
+       For example, [Lwt.join] or [Lwt.choose]. These promises only complete
+       when all or one of a set of "preceding" promises complete. The user
+       cannot complete these promises directly.
+
+
+   2. Resolvers
+
+   Resolvers are given to the user by [Lwt.wait] and [Lwt.task], and can be used
+   by the user to resolve the corresponding promises. Note that this means the
```

Comment:
FWIW, here's the E terminology:

http://www.erights.org/elib/concurrency/refmech.html

- There, a "resolved" value can be either "fulfilled" or "broken".

(the E model is more complex than the Lwt one because it also handles remote promises)